### PR TITLE
[Inference] Add CacheBlock and KV-Cache Manager

### PR DESCRIPTION
**Pull request by yuanheng-zhao**
## 📌 Checklist before creating the PR

- [x] I have created an issue for this PR for traceability
- [x] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [ ] I have added relevant tags if possible for us to better distinguish different PRs

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

#5154 

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.

This PR adds logical cache block structure, and the KV-Cache Manager which manages both the logical cache blocks and physical KV caches (tensors).

image

## 💥 Checklist before requesting a review

- [x] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [x] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [x] I have added thorough tests.
- [x] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: colossalai/inference/kv_cache/kvcache_manager.py [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+        self.max_batch_size = config.max_batch_size
+        self.max_input_length = config.max_input_length
+        self.max_output_length = config.max_output_length
+        # Cache block settings
+        self.block_size = config.block_size
+        # NOTE: `num_blocks` is not prompted, but evaluated from the maximum input/output length, and the maximum batch size
+        self.max_blocks_per_sequence = (
+            self.max_input_length + self.max_output_length + self.block_size - 1
+        ) // self.block_size
+        self.num_blocks = self.max_blocks_per_sequence * self.max_batch_size * self.beam_width
```
[the remark this answers is not in the document]
Comment (yuanheng-zhao) [reply colossalai/inference/kv_cache/kvcache_manager.py:46]: We still want to have a `max_batch_size` as we need to assign the physical caches. We're not going to infer the size of physical cache from hardware status and a ratio of gpu usage, for the evaluation might be inaccurate.

[the remark this answers is not in the document]
Comment (CjhHa1) [reply colossalai/inference/kv_cache/kvcache_manager.py:46]: a `max_batch` size is still inaccurate

File: colossalai/inference/kv_cache/kvcache_manager.py [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+            if block_id >= 0:
+                block: CacheBlock = self._cache_blocks[block_id]
+                k_ptrs.append(block.k_ptrs[layer_id])
+                v_ptrs.append(block.v_ptrs[layer_id])
+        return k_ptrs, v_ptrs
+
+    def allocate_from_block_table(
+        self, block_table: torch.Tensor, already_allocated_len: int, space_asked: int = 1
+    ) -> None:
+        """Allocate the logical cache blocks for a single sequence.
```
[the remark this answers is not in the document]
Comment (yuanheng-zhao) [reply colossalai/inference/kv_cache/kvcache_manager.py:127]: Block table shall be prepared and hold in request handler or the engine, and would be passed to model modeling.

[the remark this answers is not in the document]
Comment (CjhHa1) [reply colossalai/inference/kv_cache/kvcache_manager.py:127]: why not just let cache manager hold block table?

File: colossalai/inference/config.py [hunk 3]
```diff
@@ -1,7 +1,34 @@
...
-    3. cache_config: Configuration for initialize and manage kv cache, including `block size`, `block num`
-For the convenience of users, we provide a unified config api for that wrapped all the configs. One can easily construct a colossal_config by setting the needed configs.
-"""
+from dataclasses import dataclass
+
+import torch
+
+
+@dataclass
+class InferenceConfig:
```
[the remark this answers is not in the document]
Comment (yuanheng-zhao) [reply colossalai/inference/config.py:7]: Yeah, this file is like a placeholder for now, and will be removed as I will make tiny modifications on configs and imports when the PR https://github.com/hpcaitech/ColossalAI/pull/5149 (config) get merged in.

[the remark this answers is not in the document]
Comment (yuanheng-zhao) [reply colossalai/inference/config.py:7]: Done

File: colossalai/inference/core/config.py [hunk 4]
```diff
@@ -41,14 +42,14 @@ class InferenceConfig:
     max_seq_len: Optional[int] = None
     quant_mode: Optional[str] = None
     revision: Optional[str] = None
+    # TODO: beam search is not support for now
+    beam_width: int = 1
```
Comment (FrankLeeeee) [inline colossalai/inference/core/config.py:46]: Shouldn't this be one of the argument of `generate`?

File: colossalai/inference/core/config.py [hunk 5]
```diff
@@ -41,14 +42,14 @@ class InferenceConfig:
...
 
     def __post_init__(self):
         self._verify_args()
 
     def _verify_args(self):
         if self.gpu_utilization_rate > 1.0:
-            raise ValueError(
-                f"GPU utilization should be less than 1.0, but is set to {self.gpu_memory_utilization}."
-            )
+            raise ValueError(f"GPU utilization should be less than 1.0, but is set to {self.gpu_memory_utilization}.")
```
Comment (FrankLeeeee) [inline colossalai/inference/core/config.py:53]: Also check for negative number?

File: tests/test_infer/test_kvcache_manager.py [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+import torch
+from transformers.models.llama import LlamaConfig
+
+from colossalai.inference.kv_cache import CacheBlock, KVCacheManager
+from colossalai.logging import disable_existing_loggers
+from colossalai.testing import parameterize
+
+
+@dataclass
+class SampleInferenceConfig:
```
Comment (FrankLeeeee) [inline tests/test_infer/test_kvcache_manager.py:13]: This class and the `InferenceConfig` look confusing to me

File: tests/test_infer/test_kvcache_manager.py [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+    max_input_length = test_config["max_input_len"]
+    max_output_length = test_config["max_output_len"]
+
+    sample_config = SampleInferenceConfig(**test_config)
+    model_config = LlamaConfig(
+        hidden_size=hidden_size,
+        num_hidden_layers=num_layers,
+        num_attention_heads=num_attention_heads,
+    )
+    cache_manager = KVCacheManager(sample_config, model_config)
```
Comment (FrankLeeeee) [inline tests/test_infer/test_kvcache_manager.py:100]: The one arg of `KVCacheManager` is an instance of `InferenceConfig`, why create the `SampleInferenceConfig` again?

File: tests/test_infer/test_kvcache_manager.py (hunk 6 above)
Comment (yuanheng-zhao) [reply tests/test_infer/test_kvcache_manager.py]: `SampleInferenceConfig ` class was removed and pytest is using `InferenceConfig` at this moment.

File: tests/test_infer/test_kvcache_manager.py (hunk 7 above)
Comment (yuanheng-zhao) [reply tests/test_infer/test_kvcache_manager.py]: Done

File: colossalai/inference/core/config.py (hunk 4 above)
Comment (FrankLeeeee) [reply colossalai/inference/core/config.py:45]: Ok, then you should add its description to the docstring as well.

File: colossalai/inference/core/config.py [hunk 8]
```diff
@@ -34,21 +35,18 @@ class InferenceConfig:
...
+    beam_width: int = 1
 
     def __post_init__(self):
         self._verify_args()
 
     def _verify_args(self):
-        if self.gpu_utilization_rate > 1.0:
-            raise ValueError(
-                f"GPU utilization should be less than 1.0, but is set to {self.gpu_memory_utilization}."
-            )
```
Comment (FrankLeeeee) [inline colossalai/inference/core/config.py:52]: Why delete this?

Comment (yuanheng-zhao) [reply colossalai/inference/core/config.py:52]: We're not going to use GPU Utilization rate at this moment as the allocation of physical cache does not depend on hardware evaluation (instead, we have hard limits on maximum batch size, etc).

File: colossalai/inference/core/config.py (hunk 4 above)
Comment (yuanheng-zhao) [reply colossalai/inference/core/config.py:47]: Done

**Review by FrankLeeeee — APPROVED** [review]

**Pull request merged**